plugin/trino-sqlserver/src/main/java/io/trino/plugin/sqlserver/SqlServerClient.java

```diff
@@ -207,6 +219,32 @@
 
     private static final int MAX_SUPPORTED_TEMPORAL_PRECISION = 7;
 
+    private static final PredicatePushdownController SQLSERVER_CHARACTER_PUSHDOWN = (session, domain) -> {
+        if (domain.isNullableSingleValue()) {
+            return FULL_PUSHDOWN.apply(session, domain);
+        }
+
+        if (!domain.getValues().isDiscreteSet()) {
+            // Push down inequality predicate
+            ValueSet complement = domain.getValues().complement();
+            if (complement.isDiscreteSet()) {
+                return FULL_PUSHDOWN.apply(session, domain);
+            }
+            // Push down of range predicate for varchar/char types could lead to incorrect results
+            // when the remote database is case insensitive
+            return DISABLE_PUSHDOWN.apply(session, domain);
+        }
+
+        Domain simplifiedDomain = domain.simplify(getDomainCompactionThreshold(session));
+        if (!simplifiedDomain.getValues().isDiscreteSet()) {
+            // Domain#simplify can turn a discrete set into a range predicate
+            // Push down of range predicate for varchar/char types could lead to incorrect results
+            // when the remote database is case insensitive
+            return DISABLE_PUSHDOWN.apply(session, domain);
+        }
+        return FULL_PUSHDOWN.apply(session, simplifiedDomain);
+    };
```

Comment:
@Praveen2112 Can you please take a look if the Domain handling appears correct?

Reply:
Shouldn't we run `getDomainCompactionThreshold` check first and then check if the `complement` of given `ValueSet` is discrete ? What if we `complement` of a `ValueSet` is discrete but it has too many values so - we might compact it and then `domain.getValues().complement().isDiscreteSet` might not be valid. 

It would be nice if we have a test coverage for this edge case also.

Reply:
Thx, looks like I disable compaction threshold functionality by performing this check.
Refactored and test case added.